Add httpHeaders to httpGet liveness probe

therc opened the pull request:
Also update existing documentation and try to steer users away from 'host'.

Fixes #14713 

cc: @thockin 

Release Note:

Allow HTTP readinessProbes and livenessProbes to include a list of arbitrary HTTP headers.

therc commented:
I'm not sure what to do with defaults and validations.

lavalamp commented:
Oh, I was just wanting exactly this, neat. We're in the middle of fixing flaky tests so it may be a few days before I get to an actual review.

thockin commented on docs/user-guide/liveness/README.md:
```diff
@@ -66,11 +66,13 @@ The [http-liveness.yaml](http-liveness.yaml) demonstrates the HTTP check.
         path: /healthz
         port: 8080
+        httpHeaders:
+          X-Custom-Header: [ Awesome ]
```
what doe this being a list of values mean?

therc replied:
That's the actual type for http.Header. It results in the same header being output multiple times, with different values for each occurrence. I guess it might make sense for e.g. chains of X-Forwarded-For? Or for cookies, see http://stackoverflow.com/questions/3241326/set-more-than-one-http-header-with-the-same-name

lavalamp replied:
HTTP lets you specify headers multiple times. Generally I find this feature very annoying, but we definitely should support it.

thockin replied:
Sure we should support it, but should we support is as a multi-map (like this) or as a list of {key, value} (which feels  more natural to me and less "because that is how Go does it).

lavalamp replied:
Yeah, that's a good question. @thockin's way is probably more consistent with the rest of the API but the current way is going to be a lot more compact. I don't think there's much risk of us wanting to add other fields in here, so I guess I have an extremely minor preference for the current way. up to @thockin

thockin replied:
I think the vast majority of cases will be one key, one value.  Our API (and the documented conventions) eschew maps where the value is more than a simple primitive.

This can be a list of (key, value) or if you reallllly want to the map-like semantic a list of { key, []value}.

thockin commented on pkg/api/v1/types.go:
```diff
@@ -883,11 +883,14 @@ type HTTPGetAction struct {
 	// Scheme to use for connecting to the host.
 	// Defaults to HTTP.
 	Scheme URIScheme `json:"scheme,omitempty"`
+	// Custom HTTP headers to set in the request.
```
Can you explain the semantic if I have multiple values for a key?  Do I get the same headerline more than once?

```
httpHeaders:
    X-Custom-Thing: [foo, bar]
```

therc replied:
Yes. And it's in the HTTP RFC. I could make the API use a single value, then convert it to a list.

thockin commented on pkg/kubelet/prober/prober.go:
```diff
@@ -145,7 +145,9 @@ func (pb *prober) runProbe(p *api.Probe, pod *api.Pod, status api.PodStatus, con
 		url := formatURL(scheme, host, port, path)
-		return pb.http.Probe(url, timeout)
+		headers := p.HTTPGet.HTTPHeaders
+		glog.V(1).Infof("HTTP-Probe Headers: %v", headers)
```
This is too chatty - it will log every probe of every container.  V(4) seems correct

therc replied:
Sorry, I forgot to bump it back to 4 after some testing.

thockin commented:
For defaulting you should init a nil map to be an empty map.

For validation, I don't know that there's much to validate.  Does the HTTP spec say anything about what is allowed in headers, or max number of them or sizes?

therc commented:
For validation, the RFC allows alphanumeric plus a bunch of other ASCII characters (`!#$%&+-`, etc. but not `:,;` and others). The Go HTTP library, though, makes MIME and HTTP header canonical with a stricter subset: [A-Za-z0-9-]. I think I'll go with the latter, since it wouldn't make sense to specify characters that might result in erratic behaviour.

thockin commented:
In general, we can always expand the supported character set if needed, but
we can never make it more restrictive.

On Thu, Jan 28, 2016 at 9:58 AM, Rudi C <notifications@github.com> wrote:

> For validation, the RFC allows alphanumeric plus a bunch of other ASCII
> characters (!#$%&+-, etc. but not :,; and others). The Go HTTP library,
> though, makes MIME and HTTP header canonical with a stricter subset:
> [A-Za-z0-9-]. I think I'll go with the latter, since it wouldn't make sense
> to specify characters that might result in erratic behaviour.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20116#issuecomment-176305682>
> .
>

therc commented:
PTAL.
Verbose logging is gone. I added defaulting, validation (added to pkg/util/validation, too) and better comments about HTTP allowing repeated fields. I didn't squash the changes yet, as I think that will lose comment history.

lavalamp commented:
I looked over this and have no objections. switching reviewer to @thockin

therc commented:
Just fixing the testLabels test. It tripped over the defaulting. :-( That and rebasing generated code being painful.

therc commented:
Rebased, squashed, re-update-all-ed and with a fix for the dockertools test that tripped over differences due to defaulting.

therc commented:
BTW, regarding map vs. multi-map: HTTP headers are case-insensitive. Allowing a list of pairs would also require even more tests for case sensitivity.

thockin commented:
Regarding case testing - nothing stops a user from saying:

```
httpHeaders:
    foo-bar: value1
    Foo-Bar: value2
```

whether that is a map or a list of pairs, you have 2 distinct entries.

thockin commented:
LGTM except map

therc commented:
I pity the fool who signed up for an API change... code generation and changing a type in the middle of a PR are a nightmare, when together.
Anyway, I added a new HTTPHeader type (a string, string pair), moved the mapping from that to HTTP.Header into a separate function and added tests for that. I also updated docs with the new syntax:

```yaml
    livenessProbe:
      httpGet:
        path: /healthz
        port: 8080
        httpHeaders:
          - [ X-Custom-Header, Awesome ]
      initialDelaySeconds: 15
      timeoutSeconds: 1
```

thockin commented:
LOL.  The code generation thing HAS TO GET FIXED.  Sigh.  Taking a look
now, thanks.

On Fri, Jan 29, 2016 at 9:37 AM, Rudi C <notifications@github.com> wrote:

> I pity the fool who signed up for an API change... code generation and
> changing a type in the middle of a PR are a nightmare, when together.
> Anyway, I added a new HTTPHeader type (a string, string pair), moved the
> mapping from that to HTTP.Header into a separate function and added tests
> for that. I also updated docs with the new syntax:
>
>     livenessProbe:
>       httpGet:
>         path: /healthz
>         port: 8080
>         httpHeaders:
>           - [ X-Custom-Header, Awesome ]
>       initialDelaySeconds: 15
>       timeoutSeconds: 1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20116#issuecomment-176876779>
> .
>

thockin commented on docs/user-guide/liveness/http-liveness.yaml:
```diff
@@ -13,6 +13,8 @@ spec:
         path: /healthz
         port: 8080
+        httpHeaders:
+          - ['X-Custom-Header', 'Awesome']
```
I didn't realize this was valid YAML for initializing a struct - is it?  Can you please use the more conventional syntax:

```
httpHeaders:
  - name: X-Custom-Header
    value: Awesome
```

therc replied:
It's at the bottom of my YAML bible, [Complete idiot's introduction to YAML](https://github.com/Animosity/CraftIRC/wiki/Complete-idiot's-introduction-to-yaml). Fixed.

thockin replied:
The alternative format in that doc is for a list, but we need to
marshal/unmarshal as a struct (map in yaml).  It might work but it's
surprising at best.  Thanks :)

Ping me when all done and I'll final review ASAP.  This will make it into
1.2.

On Fri, Jan 29, 2016 at 9:54 AM, Rudi C <notifications@github.com> wrote:

> In docs/user-guide/liveness/http-liveness.yaml
> <https://github.com/kubernetes/kubernetes/pull/20116#discussion_r51291887>
> :
>
> > @@ -13,6 +13,8 @@ spec:
> >        httpGet:
> >          path: /healthz
> >          port: 8080
> > +        httpHeaders:
> > +          - ['X-Custom-Header', 'Awesome']
>
> It's at the bottom of my YAML bible, Complete idiot's introduction to YAML
> <https://github.com/Animosity/CraftIRC/wiki/Complete-idiot's-introduction-to-yaml>.
> Fixed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20116/files#r51291887>.
>

thockin commented on docs/user-guide/liveness/README.md:
```diff
@@ -66,11 +66,13 @@ The [http-liveness.yaml](http-liveness.yaml) demonstrates the HTTP check.
         path: /healthz
         port: 8080
+        httpHeaders:
+          - [ X-Custom-Header, Awesome ]
```
same comment as below on yaml syntax

thockin commented on pkg/api/v1/defaults.go:
```diff
@@ -187,6 +187,9 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
 			if obj.Scheme == "" {
 				obj.Scheme = URISchemeHTTP
 			}
+			if obj.HTTPHeaders == nil {
```
since it is not a map now, I don't think you need to have a default (a nil slice is appendable, a nil map is useless)

thockin replied:
less code is best code, so I think we can drop this and the test for it.

therc replied:
I added the test because in the old case the error afterwards, alone, was misleading, because there's no reflect.DeepDiff(). Let me check a bit.

thockin commented on pkg/kubelet/dockertools/labels_test.go:
```diff
@@ -40,10 +40,11 @@ func TestLabels(t *testing.T) {
+				Host:        "host",
+				Port:        intstr.FromInt(8080),
+				Scheme:      "scheme",
+				HTTPHeaders: nil,
```
is this needed?  nil should be the default, and it's not relevant to the test...

thockin commented on pkg/kubelet/dockertools/labels_test.go:
```diff
@@ -116,10 +117,12 @@ func TestLabels(t *testing.T) {
 	expected.PodTerminationGracePeriod = pod.Spec.TerminationGracePeriodSeconds
 	expected.PreStopHandler = container.Lifecycle.PreStop
+	// Touch up HTTPHeaders. Defaulting has converted them from nil to an empty array.
+	expected.PreStopHandler.HTTPGet.HTTPHeaders = make([]api.HTTPHeader, 0)
```
as above - I don't think this is needed is it?  Maybe this whole file can be reverted

thockin commented:
Thanks!  just a few small comments and it looks great.

lavalamp commented:
That's an excellent point. If the user uses non-canonical headers, they
could have lists that need to be merged. Therefore I now strongly favor
@thockin's approach.

On Thu, Jan 28, 2016 at 3:52 PM, Rudi C <notifications@github.com> wrote:

> BTW, regarding map vs. multi-map: HTTP headers are case-insensitive.
> Allowing a list of pairs would also require even more tests for case
> sensitivity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20116#issuecomment-176486493>
> .
>

therc commented:
Ready for review

thockin commented:
LGTM

Commit:
Add httpHeaders to httpGet liveness probe

Also update existing documentation and try to steer users away from 'host'.
Add validation.

therc commented:
Aaaarghhhh

therc commented:
Can anyone merge before it needs a rebase once again? :-)

therc commented:
Here we go again...

thockin commented:
Sorry, it's a busy project with a few heavily contended files...

therc commented:
"There is no hell, only the Kubernetes submit queue."

I'm assuming this is a flake and doesn't need a rebase.

therc commented:
@k8s-bot test this

(Does that work?)

therc commented:
k8s-bot test this issue: #20631

thockin commented:
I'll keep my eye on this one

thockin commented:
Sorry, rebase again.  It's on my screen all-day in case of any delays :)

therc commented:
Even preemptive rebasing was pointless...

thockin commented:
@k8s-oncall this has been a long-suffering PR, if there's bandwidth to merge it, that would be very gracious :)

lavalamp commented:
This is actually 2nd in the queue right now, and everything is green. I will just let the queue merge this unless I go on a manual merging spree--will keep this tab open.

therc commented:
Woo hoo! 
http://cdn.meme.am/instances2/500x/4167846.jpg
